File: ios/Sources/TaigiKeyboard/Lexicon/Trie/TrieService.swift
import Foundation

/// MARISA-trie 查詢服務
///
/// 提供前綴搜尋和完全匹配功能，用於快速查詢詞典索引。
/// 底層使用 C++ MARISA-trie library 透過 C 橋接層存取。
///
/// 支援多實例：每個 TrieService 實例管理一個獨立的 trie handle。
/// `.shared` 繼續管理 dictionary.trie。
final class TrieService: @unchecked Sendable {
    // MARK: - Properties

    static let shared = TrieService(
        fileName: "dictionary",
        fileExtension: "trie",
        logCategory: "TrieService",
    )

    private let logger: DebugLogger
    private let fileName: String
    private let fileExtension: String

    /// 用於保護 isInitialized / handle 的序列佇列
    private let stateQueue = DispatchQueue(label: "com.taigikeyboard.trie.state")
    private var _isInitialized = false
    private var _handle: trie_handle_t = -1

    // MARK: - Initialization

    init(fileName: String, fileExtension: String, logCategory: String = "TrieService") {
        self.fileName = fileName
        self.fileExtension = fileExtension
        logger = DebugLogger(category: logCategory)
    }

    deinit {
        stateQueue.sync {
            if _isInitialized, _handle >= 0 {
                trie_h_close(_handle)
            }
        }
    }

    // MARK: - State Access

    /// Read current handle atomically; returns -1 if not initialized
    private var currentHandle: trie_handle_t {
        stateQueue.sync { _isInitialized ? _handle : -1 }
    }

    // MARK: - Public API

    /// 初始化 trie（從 Bundle 載入）
    /// - Returns: 是否成功載入
    @discardableResult
    func initialize() -> Bool {
        stateQueue.sync {
            if _isInitialized {
                return true
            }

            guard let path = triePath else {
                logger.error("[INIT] Trie file not found in bundle: \(fileName).\(fileExtension)")
                return false
            }

            let h = trie_create(path)

            if h >= 0 {
                _handle = h
                _isInitialized = true
                let keyCount = trie_h_get_key_count(h)
                logger.info("[INIT] Trie loaded: \(fileName).\(fileExtension), keys=\(keyCount)")
            } else {
                logger.error("[INIT] Failed to load trie: \(fileName).\(fileExtension)")
            }

            return _isInitialized
        }
    }

    /// 前綴搜尋（回傳所有符合結果）
    /// - Parameter prefix: 搜尋前綴
    /// - Returns: 匹配的 rowid 列表
    func prefixSearch(_ prefix: String) -> [Int] {
        let h = currentHandle
        guard h >= 0 else {
            logger.warning("[SEARCH] Trie not initialized")
            return []
        }

        guard !prefix.isEmpty else {
            return []
        }

        let bufferSize = max(Int(trie_h_get_key_count(h)), 1000)
        var results = [Int32](repeating: 0, count: bufferSize)
        let count = trie_h_prefix_search(h, prefix, &results, Int32(bufferSize))

        if count > 0 {
            return results.prefix(Int(count)).map { Int($0) }
        }

        return []
    }

    /// 完全匹配查詢
    /// - Parameter key: 要查詢的 key
    /// - Returns: 匹配的 rowid 列表（一個 key 可能對應多個 rowid）
    func lookup(_ key: String) -> [Int] {
        let h = currentHandle
        guard h >= 0 else {
            logger.warning("[LOOKUP] Trie not initialized")
            return []
        }

        guard !key.isEmpty else {
            return []
        }

        let bufferSize = max(Int(trie_h_get_key_count(h)), 1000)
        var results = [Int32](repeating: 0, count: bufferSize)
        let count = trie_h_lookup(h, key, &results, Int32(bufferSize))

        if count > 0 {
            return results.prefix(Int(count)).map { Int($0) }
        }

        return []
    }

    /// 檢查是否已初始化
    var isReady: Bool {
        let h = currentHandle
        return h >= 0 && trie_h_is_loaded(h)
    }

    /// 釋放資源
    func close() {
        stateQueue.sync {
            if _isInitialized {
                trie_h_close(_handle)
                _handle = -1
                _isInitialized = false
                logger.info("[CLOSE] Trie closed: \(fileName).\(fileExtension)")
            }
        }
    }

    // MARK: - Private Methods

    /// 取得 trie 檔案路徑
    private var triePath: String? {
        ResourceBundleResolver.dictionaryBundle.path(
            forResource: fileName,
            ofType: fileExtension,
        )
    }
}

File: android/app/src/main/java/com/siansiansu/taigikeyboard/ime/dictionary/TrieService.kt
package com.siansiansu.taigikeyboard.ime.dictionary

import android.content.Context
import com.siansiansu.taigikeyboard.BuildConfig
import com.siansiansu.taigikeyboard.ime.core.logging.LoggerBackend
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import java.io.File
import java.io.FileOutputStream

/**
 * MARISA-trie lookup service backed by a C++ library via JNI.
 * Provides exact-match and prefix search over dictionary indices. Owned
 * by `CompositionRoot`; all JNI state lives in native `g_trie` so only
 * one instance is expected per process.
 */
class TrieService(
    appContext: Context,
    private val logger: LoggerBackend,
) {
    private val appContext: Context = appContext.applicationContext

    @Volatile
    private var isInitialized = false
    private val initMutex = Mutex()

    companion object {
        private const val TAG = "TrieService"
        private const val TRIE_FILE_NAME = "dictionary.trie"

        init {
            try {
                System.loadLibrary("taigi_trie")
                if (BuildConfig.DEBUG) {
                    android.util.Log.i(TAG, "[INIT] Native library loaded")
                }
            } catch (e: UnsatisfiedLinkError) {
                if (BuildConfig.DEBUG) {
                    android.util.Log.e(TAG, "[INIT] Failed to load native library", e)
                }
            }
        }
    }

    /** Copy trie asset on first run and hand it to the native loader. */
    suspend fun init(): Boolean =
        withContext(Dispatchers.IO) {
            if (isInitialized) return@withContext true

            initMutex.withLock {
                if (isInitialized) return@withLock true

                try {
                    val triePath = getTriePath(appContext)
                    val success = nativeLoad(triePath)

                    if (success) {
                        isInitialized = true
                        logger.i(TAG, "[INIT] Trie loaded, keys=${nativeGetKeyCount()}")
                    } else {
                        logger.e(TAG, "[INIT] Failed to load trie")
                    }

                    success
                } catch (e: Exception) {
                    logger.e(TAG, "[INIT] Exception during init", e)
                    false
                }
            }
        }

    /** Prefix search — returns every rowid whose key starts with [prefix]. */
    fun prefixSearch(prefix: String): IntArray {
        if (!isInitialized) {
            logger.w(TAG, "[SEARCH] Trie not initialized")
            return IntArray(0)
        }
        if (prefix.isEmpty()) return IntArray(0)

        return try {
            val bufferSize = maxOf(nativeGetKeyCount(), 1000)
            nativePrefixSearch(prefix, bufferSize)
        } catch (e: Exception) {
            logger.e(TAG, "[SEARCH] Prefix search failed", e)
            IntArray(0)
        }
    }

    /** Exact-match lookup — returns every rowid mapped to [key] (one-to-many). */
    fun lookup(key: String): IntArray {
        if (!isInitialized) {
            logger.w(TAG, "[LOOKUP] Trie not initialized")
            return IntArray(0)
        }
        if (key.isEmpty()) return IntArray(0)

        return try {
            nativeLookup(key)
        } catch (e: Exception) {
            logger.e(TAG, "[LOOKUP] Lookup failed", e)
            IntArray(0)
        }
    }

    /** Current native key count, or 0 when trie is not yet loaded. */
    fun getKeyCount(): Int = if (isInitialized) nativeGetKeyCount() else 0

    /** True once [init] has returned success. */
    val isReady: Boolean get() = isInitialized && nativeIsLoaded()

    /** Release the native trie handle. Safe to call from any thread. */
    @Synchronized
    fun close() {
        if (isInitialized) {
            nativeClose()
            isInitialized = false
            logger.i(TAG, "[CLOSE] Trie closed")
        }
    }

    /** Resolve trie file path, copying from assets when version changes. */
    private fun getTriePath(context: Context): String {
        val trieFile = File(context.filesDir, TRIE_FILE_NAME)
        val versionFile = File(context.filesDir, "trie_app_version.txt")

        val currentAppVersion = BuildConfig.VERSION_CODE
        val lastCopiedVersion =
            if (versionFile.exists()) {
                versionFile.readText().trim().toIntOrNull() ?: 0
            } else {
                0
            }

        if (currentAppVersion > lastCopiedVersion || !trieFile.exists()) {
            try {
                context.assets.open(TRIE_FILE_NAME).use { input ->
                    FileOutputStream(trieFile).use { output ->
                        input.copyTo(output)
                    }
                }
                versionFile.writeText(currentAppVersion.toString())

                logger.i(TAG, "[UPDATE] Trie updated from v$lastCopiedVersion to v$currentAppVersion")
            } catch (e: Exception) {
                logger.e(TAG, "[ERROR] Failed to copy trie from assets", e)
                throw e
            }
        }

        return trieFile.absolutePath
    }

    private external fun nativeLoad(path: String): Boolean

    private external fun nativePrefixSearch(
        prefix: String,
        limit: Int,
    ): IntArray

    private external fun nativeLookup(key: String): IntArray

    private external fun nativeGetKeyCount(): Int

    private external fun nativeIsLoaded(): Boolean

    private external fun nativeClose()
}
